lib/src/solver/incompatibility_cause.dart

```diff
@@ -43,6 +31,39 @@ abstract class IncompatibilityCause {
   String? get hint => null;
 }
 
+/// The incompatibility represents the requirement that the root package
+/// exists.
+class RootIncompatibilityCause extends IncompatibilityCause {}
+
+/// The incompatibility represents a package's dependency.
+class DependencyIncompatibilityCause extends IncompatibilityCause {
+  final PackageRange depender;
+  final PackageRange target;
+  DependencyIncompatibilityCause(this.depender, this.target);
+
+  @override
+  String? get notice {
+    final dependerDescription = depender.description;
+    if (dependerDescription is SdkDescription) {
+      final targetConstraint = target.constraint;
+      if (targetConstraint is Version) {
+        return '''
+Note: ${target.name} is pinned to version $targetConstraint by ${depender.name} from the ${dependerDescription.sdk} SDK.
+See https://dart.dev/go/version-pinning for details.
```

Comment:
Is this specific to dependencies pinned by SDK packages?
```suggestion
See https://dart.dev/go/sdk-version-pinning for details.
```